SlowGracefulShutdownCorrectnessTest.testPartitionData_whenNodesStartedShutdown_withRestart

_master_ (commit ede022c8f8dfe07fc29c8c8dd712102ab18195a7)

Failed on Sonar : http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.x-sonar/71/testReport/junit/com.hazelcast.internal.partition/SlowGracefulShutdownCorrectnessTest/testPartitionData_whenNodesStartedShutdown_withRestart_backups_0_nodes_6_shutdown_3_/

**Stacktrace**
```
java.lang.AssertionError: Partition: 58, replica: 0 data is missing on [127.0.0.1]:5710 - 3ad6a5ed-117a-4699-aeb8-7b42c720f0b9
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertNoMissingData(PartitionCorrectnessTestSupport.java:280)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndData(PartitionCorrectnessTestSupport.java:244)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.access$000(PartitionCorrectnessTestSupport.java:63)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport$AssertSizeAndDataTask.run(PartitionCorrectnessTestSupport.java:389)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1310)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1412)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndDataEventually(PartitionCorrectnessTestSupport.java:206)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndDataEventually(PartitionCorrectnessTestSupport.java:202)
	at com.hazelcast.internal.partition.AbstractGracefulShutdownCorrectnessTest.testPartitionData_whenNodesStartedShutdown_withRestart(AbstractGracefulShutdownCorrectnessTest.java:109)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionImpl.java
@@ -60,8 +60,23 @@ public class InternalPartitionImpl implements InternalPartition {
         return isMigrating;
     }
 
-    public void setMigrating(boolean isMigrating) {
-        this.isMigrating = isMigrating;
+    /**
+     * Sets migrating flag if it's not set already.
+     * @return true if migrating flag is updated, false otherwise
+     */
+    public boolean setMigrating() {
+        if (isMigrating) {
+            return false;
+        }
+        isMigrating = true;
+        return true;
+    }
+
+    /**
+     * Resets migrating flag.
+     */
+    public void resetMigrating() {
+        isMigrating = false;
     }
 
     @Override
@@ -234,7 +249,7 @@ public class InternalPartitionImpl implements InternalPartition {
         assert localReplica != null;
         this.replicas = new PartitionReplica[MAX_REPLICA_COUNT];
         this.localReplica = localReplica;
-        setMigrating(false);
+        resetMigrating();
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionStateManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionStateManager.java
@@ -313,18 +313,22 @@ public class PartitionStateManager {
         return newState;
     }
 
-    public void setMigratingFlag(int partitionId) {
+    public boolean trySetMigratingFlag(int partitionId) {
         if (logger.isFinestEnabled()) {
             logger.finest("Setting partition-migrating flag. partitionId=" + partitionId);
         }
-        partitions[partitionId].setMigrating(true);
+        return partitions[partitionId].setMigrating();
     }
 
     public void clearMigratingFlag(int partitionId) {
         if (logger.isFinestEnabled()) {
             logger.finest("Clearing partition-migrating flag. partitionId=" + partitionId);
         }
-        partitions[partitionId].setMigrating(false);
+        partitions[partitionId].resetMigrating();
+    }
+
+    public boolean isMigrating(int partitionId) {
+        return partitions[partitionId].isMigrating();
     }
 
     /** Sets the replica members for the {@code partitionId}. */

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BaseMigrationOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BaseMigrationOperation.java
@@ -216,7 +216,10 @@ abstract class BaseMigrationOperation extends AbstractPartitionOperation
                     + ". Current active migration is " + currentActiveMigration);
         }
         PartitionStateManager partitionStateManager = partitionService.getPartitionStateManager();
-        partitionStateManager.setMigratingFlag(migrationInfo.getPartitionId());
+        if (!partitionStateManager.trySetMigratingFlag(migrationInfo.getPartitionId())) {
+            throw new RetryableHazelcastException("Cannot set migrating flag, "
+                    + "probably previous migration's finalization is not completed yet.");
+        }
     }
 
     void onMigrationStart() {

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BeforePromotionOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BeforePromotionOperation.java
@@ -54,7 +54,10 @@ final class BeforePromotionOperation extends AbstractPromotionOperation {
 
         InternalPartitionServiceImpl service = getService();
         PartitionStateManager partitionStateManager = service.getPartitionStateManager();
-        partitionStateManager.setMigratingFlag(getPartitionId());
+        if (!partitionStateManager.trySetMigratingFlag(getPartitionId())) {
+            throw new IllegalStateException("Cannot set migrating flag, "
+                    + "probably previous migration's finalization is not completed yet.");
+        }
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/FinalizeMigrationOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/FinalizeMigrationOperation.java
@@ -68,8 +68,25 @@ public final class FinalizeMigrationOperation extends AbstractPartitionOperation
 
     @Override
     public void run() {
-        NodeEngineImpl nodeEngine = (NodeEngineImpl) getNodeEngine();
+        InternalPartitionServiceImpl partitionService = getService();
+        PartitionStateManager partitionStateManager = partitionService.getPartitionStateManager();
+        int partitionId = migrationInfo.getPartitionId();
 
+        if (isOldBackupReplicaOwner() && partitionStateManager.isMigrating(partitionId)) {
+            // On old backup replica, migrating flag is not set during migration.
+            // Because replica is copied from partition owner to new backup replica.
+            // Old backup owner is not notified about migration until migration is committed
+            // and completed migration is published by master.
+            // If this partition's migrating flag is set, then it means another migration
+            // is submitted to this member for the same partition and it's already executed.
+            // This finalization is now obsolete.
+            getLogger().fine("Cannot execute migration finalization, because this member was previous owner of a backup replica,"
+                    + " and a new migration operation has already superseded this finalization. "
+                    + "This operation is now obsolete. -> " + migrationInfo);
+            return;
+        }
+
+        NodeEngineImpl nodeEngine = (NodeEngineImpl) getNodeEngine();
         notifyServices(nodeEngine);
 
         if (endpoint == MigrationEndpoint.SOURCE && success) {
@@ -78,10 +95,7 @@ public final class FinalizeMigrationOperation extends AbstractPartitionOperation
             rollbackDestination();
         }
 
-        InternalPartitionServiceImpl partitionService = getService();
-        PartitionStateManager partitionStateManager = partitionService.getPartitionStateManager();
-        partitionStateManager.clearMigratingFlag(migrationInfo.getPartitionId());
-
+        partitionStateManager.clearMigratingFlag(partitionId);
         if (success) {
             nodeEngine.onPartitionMigrate(migrationInfo);
         }
@@ -100,9 +114,7 @@ public final class FinalizeMigrationOperation extends AbstractPartitionOperation
         // Old backup owner is not notified about migration until migration
         // is committed on destination. This is the only place on backup owner
         // knows replica is moved away from itself.
-        PartitionReplica source = migrationInfo.getSource();
-        if (source != null && migrationInfo.getSourceCurrentReplicaIndex() > 0
-                && source.isIdentical(nodeEngine.getLocalMember())) {
+        if (isOldBackupReplicaOwner()) {
             // execute beforeMigration on old backup before commit/rollback
             for (MigrationAwareService service : migrationAwareServices) {
                 beforeMigration(event, service);
@@ -215,6 +227,12 @@ public final class FinalizeMigrationOperation extends AbstractPartitionOperation
         } catch (Throwable e) {
             getLogger().warning("Error while finalizing migration -> " + event, e);
         }
+    }
+
+    private boolean isOldBackupReplicaOwner() {
+        PartitionReplica source = migrationInfo.getSource();
+        return source != null && migrationInfo.getSourceCurrentReplicaIndex() > 0
+                && source.isIdentical(getNodeEngine().getLocalMember());
     }
 
     @Override

--- a/hazelcast/src/test/java/com/hazelcast/concurrent/lock/LockBasicTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/concurrent/lock/LockBasicTest.java
@@ -328,7 +328,7 @@ public abstract class LockBasicTest extends HazelcastTestSupport {
         final int leaseTime = 1000;
 
         lock.lock(leaseTime, TimeUnit.MILLISECONDS);
-        partition.setMigrating(true);
+        partition.setMigrating();
 
         spawn(new Runnable() {
             @Override
@@ -338,7 +338,7 @@ public abstract class LockBasicTest extends HazelcastTestSupport {
                 } catch (InterruptedException e) {
                     e.printStackTrace();
                 }
-                partition.setMigrating(false);
+                partition.resetMigrating();
             }
         });
 

--- a/hazelcast/src/test/java/com/hazelcast/internal/partition/StaleReadDuringMigrationTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/partition/StaleReadDuringMigrationTest.java
@@ -77,7 +77,7 @@ public class StaleReadDuringMigrationTest extends HazelcastTestSupport {
         final int partitionId = 0;
         final InternalPartitionServiceImpl partitionService = (InternalPartitionServiceImpl) getPartitionService(instance);
         final InternalPartitionImpl partition = (InternalPartitionImpl) partitionService.getPartition(partitionId);
-        partition.setMigrating(true);
+        partition.setMigrating();
 
         final InternalOperationService operationService = getOperationService(instance);
         final InvocationBuilder invocationBuilder = operationService

--- a/hazelcast/src/test/java/com/hazelcast/internal/util/InvocationUtilTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/util/InvocationUtilTest.java
@@ -89,7 +89,7 @@ public class InvocationUtilTest extends HazelcastTestSupport {
         final InternalPartitionService partitionService = nodeEngineImpl.getPartitionService();
         final int randomPartitionId = (int) (Math.random() * partitionService.getPartitionCount());
         final InternalPartitionImpl partition = (InternalPartitionImpl) partitionService.getPartition(randomPartitionId);
-        partition.setMigrating(true);
+        partition.setMigrating();
 
         final String operationResponse = "operationResponse";
         final Operation operation = new LocalOperation(operationResponse)
@@ -104,7 +104,7 @@ public class InvocationUtilTest extends HazelcastTestSupport {
                 } catch (InterruptedException e) {
 
                 }
-                partition.setMigrating(false);
+                partition.resetMigrating();
             }
         });
 

